packages/angular/build/src/tools/angular/transformers/lazy-routes-transformer.ts

```diff
@@ -0,0 +1,212 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import assert from 'node:assert';
+import { relative } from 'node:path/posix';
+import ts from 'typescript';
+
+/**
+ * A transformer factory that adds a property to the lazy-loaded route object.
+ * This property is used to allow for the retrieval of the module path during SSR.
+ *
+ * @param compilerOptions The compiler options.
+ * @param compilerHost The compiler host.
+ * @returns A transformer factory.
+ *
+ * @example
+ * **Before:**
+ * ```ts
+ * const routes: Routes = [
+ *   {
+ *     path: 'lazy',
+ *     loadChildren: () => import('./lazy/lazy.module').then(m => m.LazyModule)
+ *   }
+ * ];
+ * ```
+ *
+ * **After:**
+ * ```ts
+ * const routes: Routes = [
+ *   {
+ *     path: 'lazy',
+ *     loadChildren: () => import('./lazy/lazy.module').then(m => m.LazyModule),
+ *     ...(typeof ngServerMode !== "undefined" && ngServerMode ? { ɵentryName: "./lazy/lazy.module.ts" }: {})
+ *   }
+ * ];
+ * ```
+ */
+export function lazyRoutesTransformer(
+  compilerOptions: ts.CompilerOptions,
+  compilerHost: ts.CompilerHost,
+): ts.TransformerFactory<ts.SourceFile> {
+  const moduleResolutionCache = compilerHost.getModuleResolutionCache?.();
+  assert(
+    typeof compilerOptions.basePath === 'string',
+    'compilerOptions.basePath should be a string.',
+  );
+  const basePath = compilerOptions.basePath;
+
+  return (context: ts.TransformationContext) => {
+    const factory = context.factory;
+
+    const visitor = (node: ts.Node): ts.Node => {
+      if (!ts.isObjectLiteralExpression(node)) {
+        // Not an object literal, so skip it.
+        return ts.visitEachChild(node, visitor, context);
+      }
+
+      const loadFunction = getLoadComponentOrChildrenProperty(node)?.initializer;
+      // Check if the initializer is an arrow function or a function expression
+      if (
+        !loadFunction ||
+        (!ts.isArrowFunction(loadFunction) && !ts.isFunctionExpression(loadFunction))
+      ) {
+        return ts.visitEachChild(node, visitor, context);
+      }
+
+      let callExpression: ts.CallExpression | undefined;
+      if (ts.isArrowFunction(loadFunction)) {
+        // For arrow functions, the body can be a call expression directly
+        if (ts.isCallExpression(loadFunction.body)) {
+          callExpression = loadFunction.body;
+        }
+      } else if (ts.isFunctionExpression(loadFunction)) {
+        // For function expressions, check if the body has a return statement with a call expression
+        const returnStatementExpression = loadFunction.body.statements.find(
+          ts.isReturnStatement,
+        )?.expression;
+
+        if (returnStatementExpression && ts.isCallExpression(returnStatementExpression)) {
+          callExpression = returnStatementExpression;
+        }
```

Comment:
Question: Should we look for a `return` statement in arrow functions too? I'm thinking of:

```typescript
{
  path: '',
  loadComponent: () => {
    return import('./app.component').then((m) => m.AppComponent);
  },
}
```

It's probably on the rarer side, but I can see this happening if `loadComponent` exceeds the line length limit and a formatter inserts a line break. Multiline `() => value` generally looks bad, so I wouldn't be surprised to see formatters convert this into a `() => { return value; }` which would break the optimization.

Reply:
added